filesystem/filesystem.go

```diff
@@ -176,6 +178,7 @@ type Mount struct {
 	DeviceNumber   DeviceNumber
 	Subtree        string
 	ReadOnly       bool
+	MetadataPath   string
```

Comment:
The need for this `MetadataPath` information associated with a `Mount struct` appeared during my testing. Without it, the chosen main mount would be the one that has a `Subtree` equal to `"/" + baseDirName` (if any). Problem is the `Path` associated with this `Mount` is used not only to locate fscrypt's metadata directory, but also to issue `ioctl` commands to lock or unlock for instance. So we could end up in a situation where a call to `fscrypt unlock /home` leads to an `ioctl(FS_IOC_ADD_ENCRYPTION_KEY)` on `/.fscrypt`. This is a problem in some circumstances; for instance I found myself not able to write to a file in a directory that I had previously unlocked.

This is why this patch implements the idea you suggested earlier, but with an additional field to the `Mount struct` to keep separated the notion of metadata directory from the notion of mount path. If `MetadataPath` is not empty, `BaseDir()` uses it to return the location of fscrypt's metadata directory, independently of the main mount path. If `MetadataPath` is empty, then `BaseDir()` proceeds with the main mount `Path` and `baseDirName` (the usual case when `.fscrypt` is not mounted directly).

Reply:
> the chosen main mount would be the one that has a Subtree equal to "/" + baseDirName

Isn't that the desired behavior?

> So we could end up in a situation where a call to fscrypt unlock /home leads to an ioctl(FS_IOC_ADD_ENCRYPTION_KEY) on /.fscrypt

That will work fine, since the ioctl operates on the filesystem, not on any particular file.  You have to pick some file or directory on the filesystem to execute it on.  Usually the root directory is chosen, but it could be any file or directory on the filesystem and the effect will be the same.